plugins/channel_track.go

```diff
@@ -196,7 +208,81 @@ func (p *ChannelTracker) nickCallback(b *seabird.Bot, m *irc.Message) {
 	//fmt.Printf("%s (%s) changed their name to %s\n", oldUser, p.uuids[newUser], newUser)
 }
 
-func (p *ChannelTracker) namesCallback(b *seabird.Bot, m *irc.Message) {
+func (p *ChannelTracker) modeCallback(b *seabird.Bot, m *irc.Message) {
+	// We only care about MODE messages where a specific user is
+	// changed.
+	if len(m.Params) < 3 {
+		return
+	}
+
+	logger := b.GetLogger()
+
+	channel := m.Params[0]
+	target := m.Params[2]
+
+	//change := m.Params[1]
+	//op := change[0]
+	//mode := change[1]
+
+	// Ensure we know about this user and this channel
+	u := p.LookupUser(target)
+	c := p.LookupChannel(channel)
+	if u == nil || c == nil {
+		logger.Warnf("Got MODE callback for %s on %s but we aren't tracking both", target, channel)
+		return
+	}
+
+	// Just send a WHO request and clear out the modes for this user because
+	// mode parsing is hard.
+	u.channels[channel] = make(map[rune]bool)
```

Comment:
The problem isn't with it being hard, as the comment implies, it's because it's impossible to correctly know the modes of a user when it's changed without sending a message to confirm.

`MODE #test-chan -o jsvana` would require sending a message because a user could be +v+o, but only +o would be reported, so we need to double check. If a user is +o and `MODE #test-chan +v jsvana` comes in, we *could* replace +v with +o, but it was left this way for simplicity.

`252 #test-chan jsvana userhost irc.somehost.com jsvana H@ :1 A Name` is an example of a WHOREPLY. It provides the highest mode which applies to a user for each channel you share.